### PR TITLE
Filter Mutect2 artifacts that arise from apparent-duplicate reads

### PR DESCRIPTION
**Pull request by takutosato**
Migrated from broadinstitute/gatk-protected#1109

### CODE REVIEW

**Comment by takutosato** [comment]
@fleharty back to you

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 1]
```diff
@@ -30,25 +30,25 @@
     @Argument(fullName = "max_germline_posterior", optional = true, doc = "Maximum posterior probability that an allele is a germline variant")
     public double maxGermlinePosterior = 0.025;
 
-    @Argument(fullName = "maxAltAllelesThreshold", optional = true, doc="filter variants with too many alt alleles")
+    @Argument(fullName = "maxAltAllelesThreshold", optional = true, doc = "filter variants with too many alt alleles")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Hate to do this, can you make these part of a separate commit.  It's best not to modify lines that aren't part of creating this filter.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 2]
```diff
@@ -30,25 +30,25 @@
...
     public int numAltAllelesThreshold = 1;
 
-    @Argument(fullName = "maxMedianMappingQualityDifference", optional = true, doc="filter variants for which alt reads' median mapping quality is much lower than ref reads' median mapping quality.")
+    @Argument(fullName = "maxMedianMappingQualityDifference", optional = true, doc = "filter variants for which alt reads' median mapping quality is much lower than ref reads' median mapping quality.")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 3]
```diff
@@ -30,25 +30,25 @@
...
     public int maxMedianMappingQualityDifference = 15;
 
-    @Argument(fullName = "maxMedianBaseQualityDifference", optional = true, doc="filter variants for which alt reads' median base quality is much lower than ref reads' median base quality.")
+    @Argument(fullName = "maxMedianBaseQualityDifference", optional = true, doc = "filter variants for which alt reads' median base quality is much lower than ref reads' median base quality.")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 4]
```diff
@@ -30,25 +30,25 @@
...
     public int maxMedianBaseQualityDifference = 10;
 
-    @Argument(fullName = "maxMedianClippingDifference", optional = true, doc="filter variants for which alt reads' median number of clipped bases is too high compared to the median for ref reads.")
+    @Argument(fullName = "maxMedianClippingDifference", optional = true, doc = "filter variants for which alt reads' median number of clipped bases is too high compared to the median for ref reads.")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 5]
```diff
@@ -30,25 +30,25 @@
...
     public int maxMedianClippingDifference = 1;
 
-    @Argument(fullName = "maxMedianFragmentLengthDifference", optional = true, doc="filter variants for which alt reads' median fragment length is very different from the median for ref reads.")
+    @Argument(fullName = "maxMedianFragmentLengthDifference", optional = true, doc = "filter variants for which alt reads' median fragment length is very different from the median for ref reads.")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 6]
```diff
@@ -30,25 +30,25 @@
...
     public int maxMedianFragmentLengthDifference = 10000;
 
-    @Argument(fullName = "minMedianReadPosition", optional = true, doc="filter variants for which the median position of alt alleles within reads is too near the end of reads.")
+    @Argument(fullName = "minMedianReadPosition", optional = true, doc = "filter variants for which the median position of alt alleles within reads is too near the end of reads.")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 7]
```diff
@@ -30,25 +30,25 @@
...
     public int minMedianReadPosition = 5;
 
-    @Argument(fullName = "maxEventsInHaplotype", optional = true, doc="Variants coming from a haplotype with more than this many events are filtered")
+    @Argument(fullName = "maxEventsInHaplotype", optional = true, doc = "Variants coming from a haplotype with more than this many events are filtered")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java [hunk 8]
```diff
@@ -57,7 +57,10 @@
     @Argument(shortName = "strand_af", fullName = "strandArtifactAlleleFraction", optional = true, doc = "Only filter a variant if the MAP estimate of allele fraction given artifact is below this number")
     public double STRAND_ARTIFACT_ALLELE_FRACTION_THRESHOLD = 0.01;
 
-    @Argument(shortName = "contaminationTable", fullName = "contaminationTable", optional = true, doc="Table containing contamination information.")
+    @Argument(shortName = "contaminationTable", fullName = "contaminationTable", optional = true, doc = "Table containing contamination information.")
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: Separate commit.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+ * Normally, they are marked as duplicates by Picard MarkDuplicates. But when we use unique molecular identifiers (UMIs),
+ * these apparent PCR-duplicate reads may get different UMIs, and hence these reads are regarded independent evidence.
+ *
+ * Although these reads have different UMIs, we suspect that they really are PCR-duplicates, for two reasons:
+ * 1) these sites are false positives, and
+ * 2) with hybrid selection, it's highly unlikely that we get multiple fragments with identical start and end positions.
+ *
+ * We don't understand why these apparent PCR-duplicates get different UMIs; investigation is underway.
+ *
+ * We filter the variant if the count is lower than a user-specified threshold.
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: I don't think this is true, is it?  We want to filter the variant if the ratio of the count of duplicate alt reads to total count of alt reads, right?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java [hunk 10]
```diff
@@ -207,10 +208,27 @@ private void applyEventDistanceFilters(final VariantContext vc, final Collection
...
+    private void applyDuplicatedAltReadFilter(final M2FiltersArgumentCollection MTFAC, final VariantContext vc, final Collection<String> filters) {
+        Genotype tumorGenotype = vc.getGenotype(tumorSample);
+
+        if (!tumorGenotype.hasExtendedAttribute(UniqueAltReadCount.UNIQUE_ALT_READ_SET_COUNT_KEY)) {
+            return;
+        }
+
+        final int uniqueReadSetCount = GATKProtectedVariantContextUtils.getAttributeAsInt(tumorGenotype, UniqueAltReadCount.UNIQUE_ALT_READ_SET_COUNT_KEY, -1);
+
+        if (uniqueReadSetCount <= MTFAC.uniqueAltReadCount) {
```
Comment (fleharty) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java]: I don't think this is doing exactly what we want.  Can we make a change to this so that it checks to see if ...
uniqueReadSetCount / totalAltReads < MTFAC.fractionAltReadCount
where fractionAltReadCount would be the user input?

Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java]: I have reservations about this. Remember that nice separation of FPs and TPs was done with a count, not fraction. A fraction may very well separate them nicely but I'd have to go back to the data and that's extra work (maybe 1-2 hours). Plus I tried to come up with a reason why using a fraction was better but I couldn't - to me thinking of the filter in terms of counts is easier to conceptualize

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+ * have the profile where the evidence for alternate allele comes solely from one, two, or three sets of apparent PCR-duplicates.
+ * Reads in such a set have the same read-start and mate-end position i.e. they come from the same original insert.
+ * Normally, they are marked as duplicates by Picard MarkDuplicates. But when we use unique molecular identifiers (UMIs),
+ * these apparent PCR-duplicate reads may get different UMIs, and hence these reads are regarded independent evidence.
+ *
+ * Although these reads have different UMIs, we suspect that they really are PCR-duplicates, for two reasons:
+ * 1) these sites are false positives, and
+ * 2) with hybrid selection, it's highly unlikely that we get multiple fragments with identical start and end positions.
+ *
+ * We don't understand why these apparent PCR-duplicates get different UMIs; investigation is underway.
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/UniqueAltReadCount.java]: pasted the comment with a typo fix

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/UniqueAltReadCountUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+import org.broadinstitute.hellbender.utils.read.ArtificialReadUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+import java.io.IOException;
+import java.util.*;
+
+public class UniqueAltReadCountUnitTest {
+    final String sampleName = "Mark";
```
Comment (takutosato) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/mutect/UniqueAltReadCountUnitTest.java]: yup it's staying

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java (hunk 1 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: markkkk

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java [hunk 13]
```diff
@@ -207,10 +208,27 @@ private void applyEventDistanceFilters(final VariantContext vc, final Collection
         }
     }
 
+    // This filter checks for the case in which PCR-duplicates with unique UMIs (which we assume is caused by barcode swap)
```
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect2FilteringEngine.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java (hunk 1 above)
Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/M2FiltersArgumentCollection.java]: I tried to separate them. But I didn't do it quite right. I already spent too much time on this, I think it's ok for them to be in the same commit. It's been done before in the Mutect code base. I appreciate your enforcing good version control practice though.

**Comment by takutosato** [comment]
Back to you Mark (post-review changes are in the commit "review comments")

**Comment by fleharty** [comment]
@takutosato 
I know it's annoying, but I would like for you to put the edits with the spaces in a separate commit.  Then I think this will be ready to merge!

**Comment by fleharty** [comment]
Per our discussion with @davidbenjamin I think this looks good.
Can we get this to pass Travis before I give the thumbs up?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/FilterMutectCalls.java [hunk 14]
```diff
@@ -88,6 +88,8 @@ public void onTraversalStart() {
         GATKVCFConstants.MUTECT_FILTER_NAMES.stream().map(GATKVCFHeaderLines::getFilterLine).forEach(headerLines::add);
 
         headerLines.addAll(getDefaultToolVCFHeaderLines());
+        headerLines.add(new VCFFilterHeaderLine(Mutect2FilteringEngine.DUPLICATED_EVIDENCE_FILTER_NAME, "evidence for alt allele is overrepresented by apparent duplicates"));
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/FilterMutectCalls.java]: @takutosato In light of PR #3172, you should move `DUPLICATED_EVIDENCE_FILTER_NAME` to the M2 filters sections in `GATKVCFConstants` and add it to the `MUTECT_FILTER_NAMES` list.  Then, in the `M2-related filters` section of the `static` block in `GATKVCFHeaderLine` you should `addFilterLine` for this new filter.

Comment (takutosato) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/FilterMutectCalls.java]: done. I think this will fix the failed tests.
